Postprocessors for measuring finite volume quantity conservations

GiudGiud opened the pull request:
Add InternalVolumetricFlowRate for internal boundaries.
Refs #16099, would benefit from #16585

This shows the current state of quantity conservation in INSFV. 
-Mass is conserved in all instances.
-Energy is almost perfectly conserved. It's unclear what are the discrepancies. My current best guesses:
1. For internal boundaries: getInternalValue of the velocity using the extended stencil and the advection kernel is using average interpolation.
2. For external boundaries: the material property (energy) should be extrapolated to the boundary. Right now we take the centroid value

-Momentum looks good in straight channel tests. We do expect some error from the advecting velocity not being interpolated the same way in the kernels and the postprocessor. 

Note that VolumetricFlowRate, for internal boundaries, does not have access to Rhie Chow interpolation of velocity currently, so it can't show conservation. This will improve when #16585 is fixed. 

Note the trimesh example does not converge with two term boundary expansions, so it's shifted to a single term.

Thoughts on InternalVolumetricFlowRate vs InterfaceVolumetricFlowRate ? Interface is more like other moose objects, internal is more explicit. Interface is often used for internal boundaries, where one variable would stop being defined though

GiudGiud committed:
Add advected quantity to VolumetricFlowRate postprocessor
Add BoundaryVolumetricFlowRate for external boundaries
Refs #16099, would benefit from #16585

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/664918/) on 5bc071f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16767/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16767/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2f98a2971888afb9db9d489c85f300d9cb690c02`

GiudGiud committed:
Add csv golded outputs to conservation test
Fix style

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/665061/) on f776cde wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16767/site)

This comment will be updated on new commits.

GiudGiud commented:
@lindsayad please have a look to let me know if there's a better way to get the FV variable, since these postprocs usually only have the VariableValue.

GiudGiud committed:
Add support for FV variables in diffusive flux user objects and postprocessors

GiudGiud committed:
Rename flow PP

GiudGiud committed:
Add FV support for InterfaceIntegralVariableValuePostprocessor

GiudGiud committed:
Rework advective PP using Coupleable interface instead of SubProblem

GiudGiud committed:
Clang formatting previous commits

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/665938/) on d239a73 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16767/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16767/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2f98a2971888afb9db9d489c85f300d9cb690c02`

GiudGiud committed:
Dubious clang-format changes
Change documentation name

GiudGiud committed:
Fix documentation, test failures because of material ADtype, test failures because of tri-mesh in FV
Cleanup of userobjects/layered_side_* tests

GiudGiud commented:
For some reason, the gold exodus files i generate on my WSL create a diff on civet.

lindsayad commented:
Why did all these test targets get cancelled? With respect to the exodiff...can you increase the solve tolerance, e.g. solve tighter? That is our usual strategy (preferred over loosening comparison tolerances (not that you are doing that)).

lindsayad commented:
So it sounds like in order to close #16099 we will have to create our `UserObject` or whatever design pattern we choose in order to incorporate Rhie-Chow interpolations, e.g. we must be able to conduct these interpolations in contexts other than residual objects...?

GiudGiud commented:
Essentially yes. And that object should be aware of pressure, material properties, but also momentum sources (kernels? but should we use something else?)
The Rhie Chow interpolation for velocity is used everywhere except in the viscosity diffusion kernels. I wonder if we should change that too.

lindsayad commented:
> The Rhie Chow interpolation for velocity is used everywhere except in the viscosity diffusion kernels.

It is not used in the `_adv_quant` members in the advection kernels, which I believe is consistent with the methods Moukalled outlines.

lindsayad commented:
Let me know when this is ready for review

GiudGiud commented:
Ok this passes on HPC too. Only fails on CIVET. I ll look into tightening the tolerances.

GiudGiud committed:
Tighten test tolerances, update golded exodus

GiudGiud commented:
I think I convinced myself we should be using RC for the diffusion kernels.
Looking at eq. 15.13 we can see that diffusion and convection are treated together, and act on these u_f, which are the Rhie Chow interpolated velocities

lindsayad commented:
> I think I convinced myself we should be using RC for the diffusion kernels.
> Looking at eq. 15.13 we can see that diffusion and convection are treated together, and act on these u_f, which are the Rhie Chow interpolated velocities

Where would you even use this for the viscosity term? In computing the cell centroid gradients? This just doesn't feel right

lindsayad commented:
The only places lower case subscripts should appear is in mass flow rates across faces and in computation of the cell centroid gradients. And then we use 9.33 to obtain face gradients for the viscous term computation

GiudGiud commented:
Maybe. It's not clear in my mind tbh. 
The end goal is to get face gradients for the diffusion kernels. These are expressed using u_C and u_f, and u_f should use RC. Does that make sense?

lindsayad commented:
For an orthogonal grid, \nabla \phi_f depends only on neighboring cell centroid values, \phi_C and \phi_F, where F is capitalized.  It is only for non-orthogonal grids that you introduce equations which involve \phi_f.

GiudGiud commented:
True, phi_F is the neighbor value.
So nothing to do in 9.33. That only leaves the computation of the centroid gradient.

Ok about reducing to the correct form for orthogonal grid, but we should consider that the face gradient would be more consistent with the face value if we used RC to compute the face values in it. 
Right now, if you take u_C + d_Cf grad u_C you're dont get u_f (duh)

lindsayad commented:
Well it depends what u_f you're talking about right? Is it one we're interpolating with the `velocity_interp_method` or is it the one we're computing with the `advected_interp_method`?

lindsayad commented:
I suppose you could always just try it and see whether mass and momentum are conserved, and whether you maintain second order convergence via MMS when refining the mesh.

GiudGiud commented:
True.
I think I might have gotten confused too. RC is used for the mdot term in the a coeffs, which feels like it should be the advected term (it's momentum advection after all) but really it's the velocity term with regards to our naming convention for interpolation. 

Test failures are unrelated I think. This should be ready for review

lindsayad commented:
It would be nice to have a detailed fully-coupled Rhie-Chow reference. Have you encountered one? I am not 100% confident that what we have currently is correct. I am willing to be convinced that what we have is wrong, if you lead me through a trail of equations.

On another note, if we conserve mass and momentum and we demonstrate second order convergence, is it fair to say we are right? I'm not asking rhetorically

lindsayad commented:
> Test failures are unrelated I think. Some csv results arent great still. Will look at it more, but can use a review

Ok will probably do tomorrow.

lindsayad commented on framework/src/userobject/SideIntegralVariableUserObject.C:
```diff
@@ -28,13 +28,27 @@ SideIntegralVariableUserObject::SideIntegralVariableUserObject(const InputParame
+    /// We should be at the edge of the domain
+    const FaceInfo * const fi = _mesh.faceInfo(_current_elem, _current_side);
+    mooseAssert(fi, "We should have a face info");
+    if (!fi->isBoundary())
+      mooseError("SideIntegralVariableUserObject should only be used at boundaries");
```
Let's turn this into an assert. For qp methods, I'd like for things to be as fast as possible in optimized modes.

lindsayad reviewed (changes requested):
I just reviewed the framework changes. Once we've ironed out any discussion on that, I'll go on and review the module level changes

GiudGiud commented on framework/src/postprocessors/InterfaceIntegralVariableValuePostprocessor.C:
```diff
@@ -46,13 +46,41 @@ InterfaceIntegralVariableValuePostprocessor::InterfaceIntegralVariableValuePostp
 {
   addMooseVariableDependency(&mooseVariableField());
+
+  /// Check that the secondary variable is also a finite volume variable
+  if (_fv)
+    if (!getFieldVar("neighbor_variable", 0)->isFV())
+      mooseError("For the InterfaceIntegralVariableValuePostprocessor, if variable "
+                 "is a finite volume variable, so should be neighbor_variable.");
```
we sure should. Adding it

GiudGiud commented on framework/src/postprocessors/InterfaceIntegralVariableValuePostprocessor.C:
```diff
@@ -46,13 +46,41 @@ InterfaceIntegralVariableValuePostprocessor::InterfaceIntegralVariableValuePostp
+
+    /// Get primary variable on the interface
+    const Real u =
+        MetaPhysicL::raw_value(_fv_variable->getInternalFaceValue(neighbor, *fi, _u[_qp]));
```
I see what you mean. I was under the impression that 'Side'UO/PP were for boundaries, and 'Interface'UO were for internal faces.
I m thinking let's assume it's a boundary when two different variables are used, and let's assume it's not an internal interface when the same variable is used twice

lindsayad replied:
I would say we have 5 types of "geometric"-based user objects:
- `ElementUserObject`
- `SideUserObject` (boundaries)
- `InternalSideUserObject`
- `InterfaceUserObject`
- `NodalUserObject`

GiudGiud replied:
ok I might have missed the InternalSide ones. I ll have a look to make sure they work for FV then

GiudGiud replied:
ok there's just very few in the framework, and none as far as I can find that currently deal with variables. 
The layeredSide stuff could make the list too

GiudGiud commented on framework/src/postprocessors/InterfaceIntegralVariableValuePostprocessor.C:
```diff
@@ -46,13 +46,41 @@ InterfaceIntegralVariableValuePostprocessor::InterfaceIntegralVariableValuePostp
+        MetaPhysicL::raw_value(_fv_variable->getInternalFaceValue(neighbor, *fi, _u[_qp]));
+
+    /// Get secondary variable on the interface
+    const Real u_neighbor = MetaPhysicL::raw_value(
+        dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar("neighbor_variable", 0))
+            ->getInternalFaceValue(neighbor, *fi, _u_neighbor[_qp]));
```
good idea. I ll try to keep that in mind for future objects with action at each _qp

GiudGiud committed:
Address @lindsayad review:
reduce dynamic cast use, use mooseAssert in computeQpRes, reference mat-rework issue
Use interpolation for advected variables at boundaries, to be consistent with advection kernels.
Fix checks for advected quantity inputs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/670601/) on 95ae04d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16767/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16767/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2f98a2971888afb9db9d489c85f300d9cb690c02`

GiudGiud commented on modules/navier_stokes/src/postprocessors/VolumetricFlowRate.C:
```diff
@@ -16,23 +16,91 @@ InputParameters
+    /// Get face value for velocity
+    const auto & vx_face = !getFieldVar("vel_x", 0)
+                               ? _vel_x[_qp]
+                               : MetaPhysicL::raw_value(_fv_vel_x->getBoundaryFaceValue(*fi));
```
in case you're wondering this whole shenanigan is to capture the case where a user would set "vel_x = 2" instead of a variable

GiudGiud commented:
Just realized I forgot a level of constness. 
Will address the test failre and the merge conflict then this will be ready for review again

GiudGiud committed:
Use interpolation method in flowrate postprocs, update gold. Upwind has issue with dirichlet

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/671950/) on 612f5b2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16767/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16767/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b28b88dfcc3512ec6b067f428c14e805bcf28a1c`

lindsayad commented on framework/include/postprocessors/InterfaceIntegralVariableValuePostprocessor.h:
```diff
@@ -39,6 +39,11 @@ class InterfaceIntegralVariableValuePostprocessor : public InterfaceIntegralPost
   /// the type of interface value we want to compute
   const MooseEnum _interface_value_type;
+  /// the fv variable for the neighbor variable
+  const MooseVariableFV<Real> * _neighbor_fv_variable;
```
Let's do the full deal: `const MooseVariableFV<Real> * const _neighbor_fv_variable` in case some evil developer comes along in the future and tries to change what is pointed to

lindsayad commented on framework/include/userobject/SideIntegralVariableUserObject.h:
```diff
@@ -39,5 +39,6 @@ class SideIntegralVariableUserObject : public SideIntegralUserObject,
   /// Holds the solution gradient at the current quadrature points
   const VariableGradient & _grad_u;
+  /// Whether FV variables are used
+  bool _fv;
```
`const`?

lindsayad commented on framework/src/postprocessors/InterfaceIntegralVariableValuePostprocessor.C:
```diff
@@ -46,13 +46,52 @@ InterfaceIntegralVariableValuePostprocessor::InterfaceIntegralVariableValuePostp
+  /// Primary and secondary variable should both be finite volume variables or both not be
+  if ((_fv && !getFieldVar("neighbor_variable", 0)->isFV()) ||
+      (!_fv && getFieldVar("neighbor_variable", 0)->isFV()))
+    mooseError("For the InterfaceIntegralVariableValuePostprocessor, if variable "
+               "is a finite volume variable, so should be neighbor_variable.");
```
You might just say that the variable types of `variable` and `neighbor_variable` must match or use some other language that isn't finite volume of finite element specific

lindsayad commented on framework/src/postprocessors/InterfaceIntegralVariableValuePostprocessor.C:
```diff
@@ -46,13 +46,52 @@ InterfaceIntegralVariableValuePostprocessor::InterfaceIntegralVariableValuePostp
+    /// Get FaceInfo from the mesh
+    const FaceInfo * const fi = _mesh.faceInfo(_current_elem, _current_side);
+    mooseAssert(fi, "We should have a face info");
+    const Elem * neighbor = _current_elem->neighbor_ptr(_current_side);
```
Not necessary to change, but I always like to mark my pointers as `const` wherever it makes sense to do so

lindsayad commented on framework/src/postprocessors/InterfaceIntegralVariableValuePostprocessor.C:
```diff
@@ -46,13 +46,52 @@ InterfaceIntegralVariableValuePostprocessor::InterfaceIntegralVariableValuePostp
+#ifdef MOOSE_GLOBAL_AD_INDEXING
+  if (_fv)
+  {
+    /// Get FaceInfo from the mesh
```
When not writing data member doxygen, our typical style is to use just two slashes `//` for comments

GiudGiud replied:
makes sense. Not sure why I changed

lindsayad commented on framework/src/postprocessors/InterfaceIntegralVariableValuePostprocessor.C:
```diff
@@ -46,13 +46,52 @@ InterfaceIntegralVariableValuePostprocessor::InterfaceIntegralVariableValuePostp
+      u = MetaPhysicL::raw_value(_fv_variable->getBoundaryFaceValue(*fi));
+      u_neighbor = MetaPhysicL::raw_value(_neighbor_fv_variable->getBoundaryFaceValue(*fi));
+    }
+    /// If only one variable is specified, assume this is an internal interface
```
`//`

lindsayad commented on framework/src/postprocessors/InterfaceIntegralVariableValuePostprocessor.C:
```diff
@@ -46,13 +46,52 @@ InterfaceIntegralVariableValuePostprocessor::InterfaceIntegralVariableValuePostp
+    mooseAssert(fi, "We should have a face info");
+    const Elem * neighbor = _current_elem->neighbor_ptr(_current_side);
+
+    /// If both variables are different, assume this is a boundary for both variables
+    Real u, u_neighbor;
+    if (_fv_variable != _neighbor_fv_variable)
+    {
+      u = MetaPhysicL::raw_value(_fv_variable->getBoundaryFaceValue(*fi));
+      u_neighbor = MetaPhysicL::raw_value(_neighbor_fv_variable->getBoundaryFaceValue(*fi));
+    }
+    /// If only one variable is specified, assume this is an internal interface
+    // FIXME Make sure getInternalFaceValue uses the right interpolation method, see #16585
+    else
+    {
+      u = MetaPhysicL::raw_value(_fv_variable->getInternalFaceValue(neighbor, *fi, _u[_qp]));
+      u_neighbor = MetaPhysicL::raw_value(
+          _neighbor_fv_variable->getInternalFaceValue(neighbor, *fi, _u_neighbor[_qp]));
+    }
```
I like conceptually what you have done here

lindsayad commented on framework/src/postprocessors/SideIntegralVariablePostprocessor.C:
```diff
@@ -48,6 +48,8 @@ SideIntegralVariablePostprocessor::computeQpIntegral()
   {
     const FaceInfo * const fi = _mesh.faceInfo(_current_elem, _current_side);
     mooseAssert(fi, "We should have an fi");
+    mooseAssert(fi->isBoundary(),
+                "SideIntegralVariablePostprocessor should only be used at boundaries");
```
Hmm I don't really agree with this assertion. I think we should be able to run this on internal mesh sidesets as long as those sidesets represent a boundary for where that variable lives.

GiudGiud replied:
ok deleting, getBoundaryFaceValue will catch the issue if not used at a boundary for that variable anyway.

Rewording it to be more friendly since it becomes a user-facing assert

lindsayad commented on framework/src/userobject/SideIntegralVariableUserObject.C:
```diff
@@ -28,13 +28,27 @@ SideIntegralVariableUserObject::SideIntegralVariableUserObject(const InputParame
+    /// We should be at the edge of the domain
+    const FaceInfo * const fi = _mesh.faceInfo(_current_elem, _current_side);
+    mooseAssert(fi, "We should have a face info");
+    mooseAssert(fi->isBoundary(),
+                "SideIntegralVariableUserObject should only be used at boundaries");
```
As above, I think we should be able to run this on internal mesh boundaries that represent boundaries for the variable of interest

lindsayad commented on modules/navier_stokes/doc/content/source/postprocessors/InternalVolumetricFlowRate.md:
```diff
@@ -0,0 +1,7 @@
+!syntax description /Postprocessors/InternalVolumetricFlowRate
```
I know there were bad examples in our Navier-Stokes documentation before, but we need a header like
```
# InternalVolumetricFlowRate
```
at the top of every markdown file or else our source pages will look messed up. (Instead of a link that says `InternalVolumetricFlowRate` it will say something like `Input Parameters`)

lindsayad commented on modules/navier_stokes/include/postprocessors/InternalVolumetricFlowRate.h:
```diff
@@ -0,0 +1,52 @@
+#include "InterfaceIntegralPostprocessor.h"
+#include "FVUtils.h"
+
+// Forward Declarations
```
can delete

lindsayad commented on modules/navier_stokes/include/postprocessors/InternalVolumetricFlowRate.h:
```diff
@@ -0,0 +1,52 @@
+// Forward Declarations
+
+/**
+ * This postprocessor computes the volumetric flow rate through a boundary.
```
```suggestion
 * This postprocessor computes the volumetric flow rate through an internal boundary.
```

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+    /// Get face value for velocity
+    // FIXME Make sure getInternalFaceValue uses the right interpolation method, see #16585
+    const auto & vx_face =
+        !getFieldVar("vel_x", 0)
```
We can avoid calling this for every qp. In your constructor, you can initialize a boolean member like `_is_coupled_vel_x` with the result of `isCoupled("vel_x")`. Let's do this for all the velocity components

lindsayad replied:
```c++
InternalVolumetricFlowRate::InternalVolumetricFlowRate(const InputParameters & parameters)
  : InterfaceIntegralPostprocessor(parameters),
_is_coupled_vel_x(isCoupled("vel_x")),
_is_coupled_vel_y(isCoupled("vel_y")),
_is_coupled_vel_z(isCoupled("vel_z")),
```

lindsayad replied:
At some point, we will really want to implementing something such that we can avoid this ternary. Maybe it means that when we loop over internal faces in our postprocessor execution loop we populate a `MooseArray` like `_face_sln` in `MooseVariableFV` with the results of `getInternalFaceValue`. @rwcarlsen and I have talked about doing this for other quantities.

lindsayad replied:
I'm not going to make you do that here though. That's beyond the scope of this PR

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+      if (!parameters().isParamSetByUser("advected_interp_method"))
+      {
+        advected_quantity =
+            !getFieldVar("advected_variable", 0)
```
If the user set the `advected_variable` parameter, then shouldn't we be guaranteed to have `_fv_advected_variable`? I feel like we should be able to avoid the ternary here.

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+
+    /// Compute the advected quantity on the face
+    Real advected_quantity;
+    if (parameters().isParamSetByUser("advected_variable"))
```
Let's also make this into a boolean data member that we initialize in the object constructor

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+    Real advected_quantity;
+    if (parameters().isParamSetByUser("advected_variable"))
+    {
+      // If user did not request an interpolation method, use what the kernels are
+      // most likely to use
+      if (!parameters().isParamSetByUser("advected_interp_method"))
+      {
+        advected_quantity =
+            !getFieldVar("advected_variable", 0)
+                ? _advected_variable[_qp]
+                : MetaPhysicL::raw_value(_fv_advected_variable->getInternalFaceValue(
+                      neighbor, *fi, _advected_variable[_qp]));
```
I feel like we should always use `advected_interp_method` because that is what our kernels always use (whether it be explicitly specified in the input file or defaulted to `upwind`)

GiudGiud replied:
ok made the switch. Makes sense for an ns module postproc since we are using those interpolation everywhere

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+      else
+      {
+        /// Get neighbor value
+        const auto & advected_variable_neighbor =
+            !getFieldVar("advected_variable", 0)
+                ? _advected_variable[_qp]
+                : MetaPhysicL::raw_value(_fv_advected_variable->getNeighborValue(
+                      neighbor, *fi, _advected_variable[_qp]));
```
Again I'm hoping that we're guaranteed to have `_fv_advected_variable` here such that we can avoid the ternary

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+                               true);
+      }
+    }
+    else if (parameters().isParamSetByUser("advected_mat_prop"))
```
Make boolean data member initialized in constructor

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+  }
+  else
+#endif
+      if (parameters().isParamSetByUser("advected_variable"))
+    return _advected_variable[_qp] * RealVectorValue(_vel_x[_qp], _vel_y[_qp], _vel_z[_qp]) *
+           _normals[_qp];
+  else if (parameters().isParamSetByUser("advected_mat_prop"))
+    return MetaPhysicL::raw_value(_advected_material_property[_qp]) *
+           RealVectorValue(_vel_x[_qp], _vel_y[_qp], _vel_z[_qp]) * _normals[_qp];
+  else
+    return RealVectorValue(_vel_x[_qp], _vel_y[_qp], _vel_z[_qp]) * _normals[_qp];
```
These should all be finite element cases right? If so, I think we should wrap this whole section in braces. That should make sense regardless of whether `MOOSE_GLOBAL_AD_INDEXING` is defined, I think.

GiudGiud replied:
yup.
Good call this fixes the clang formatting abuse of power

lindsayad commented on modules/navier_stokes/src/postprocessors/VolumetricFlowRate.C:
```diff
@@ -16,23 +17,125 @@ InputParameters
 VolumetricFlowRate::computeQpIntegral()
 {
-  return RealVectorValue(_vel_x[_qp], _vel_y[_qp], _vel_z[_qp]) * _normals[_qp];
+#ifdef MOOSE_GLOBAL_AD_INDEXING
```
All my comments from `InternalVolumetricFlowRate` apply here as well

lindsayad commented on test/tests/userobjects/layered_side_integral/tests:
```diff
@@ -29,6 +29,13 @@
+    type = 'Exodiff'
+    input = 'layered_side_flux_average_fv.i'
+    exodiff = 'layered_side_flux_average_fv_out.e'
+    requirement = "MOOSE shall allow taking averages of the diffusive flux of finite volume variables along a coordinate axis in layers on a boundary"
```
For SQA reporting I think these requirements are supposed to be complete sentences, so go ahead and make sure there are periods at the end.

GiudGiud replied:
ok changed for the other tests in the file too, I m glad we will be back in compliance.

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/tests:
```diff
@@ -0,0 +1,86 @@
+    input = conservation_INSFE.i
+    csvdiff = straight_INSFE.csv
+    abs_zero = 1e-8
+    requirement = 'The system shall be able to compute mass and momentum flow rates at internal and external boundaries of a straight channel with a finite element incompressible Navier Stokes model.'
```
I am curious whether you observe conservation for FE?

GiudGiud replied:
Nah not at all. Like it's abysmal with such a coarse mesh too

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/tests:
```diff
@@ -0,0 +1,86 @@
+      input = conservation_INSFV.i
+      csvdiff = diverging_INSFV_quad_xy.csv
+      abs_zero = 1e-8
+      detail = 'with a quadrilateral mesh in XY geometry, with mass flow measured using either a variable or material property.'
```
for these details, I believe you want a comma instead of a period. Other than that these are fantastic requirements!

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/tests:
```diff
@@ -0,0 +1,86 @@
+      input = conservation_INSFV.i
+      csvdiff = diverging_INSFV_quad_xy_noslip.csv
+      abs_zero = 1e-8
+      detail = 'with no-slip boundary conditions, for which momentum and energy will be dissipated at the wall.'
```
```suggestion
      detail = 'and with no-slip boundary conditions, for which momentum and energy will be dissipated at the wall.'
```

lindsayad reviewed (changes requested):
Really nice work. Please let me know if you have questions about my comments. I love your `postprocessors` `tests` spec. Very detailed requirements which give me confidence that we are doing a good job conserving things!

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+
+    const auto & vy_face =
+        !getFieldVar("vel_y", 0)
+            ? _vel_y[_qp]
```
```suggestion
            ? Real(0)
```

lindsayad replied:
This one is not required, so the ternary is ok I think, but I would make it even more transparent about what is happening, e.g. the explicit assignment to `Real(0)`. (I think that will compile fine, but not 100% sure)

GiudGiud replied:
see above

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+
+    const auto & vz_face =
+        !getFieldVar("vel_z", 0)
+            ? _vel_z[_qp]
```
```suggestion
            ? Real(0)
```

lindsayad reviewed (changes requested):
I was noodling on this a bit more, and I would like to see ternaries removed wherever we can, and where we can't let's explicitly initialize to the default 0 value. (Although if you want to fight back on the `Real(0)` vs `_vel_i[_qp]` I might yield...it's just that it's not immediately obvious what `_vel_i[_qp]` means when the variable is not coupled unless you know how default values work (which I guess anyone looking seriously in this code should know))

GiudGiud commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -0,0 +1,162 @@
+    // FIXME Make sure getInternalFaceValue uses the right interpolation method, see #16585
+    const auto & vx_face =
+        !getFieldVar("vel_x", 0)
+            ? _vel_x[_qp]
```
but what if the user sets vel_x = 4 in their input

lindsayad replied:
Ah, ya got me. Keep as is!

GiudGiud committed:
Address @lindsayad review:
- code style compliance
- const pointers to fend off pointer changes
- SQA formatting for tests
- use Neumann BC when using an upwind to have the correct inflow of energy (advected mat property)

GiudGiud commented:
Comments should be addressed now. 
Unless there is a way (and a desire) to forbid setting vel_x = 2 or advected_variable = 3, I think the clean way to remove all these ternaries is some sort of coupledFVValue which has the side/boundary face values as you mentioned. That way we dont need to have the fv variable handy, and dont risk using a null pointer when users want to use a constant value.

GiudGiud commented:
As you can see in the csv, the inlet energy flow is still not '300' as it should be, but depends on the numerical scheme used. Further in the channel we get the right value, especially when the flow is close to being fully developed again. I expect this is due to not using the face value for the material property, the heat production is happening in the first cell too so the temperature has already increased.

A lot of the other values are exactly where we want them to be looking at the straight channel case.

lindsayad commented on framework/src/variables/MooseVariableFV.C:
```diff
@@ -648,7 +648,7 @@ template <typename OutputType>
 MooseVariableFV<OutputType>::getBoundaryFaceValue(const FaceInfo & fi) const
 {
-  mooseAssert(!isInternalFace(fi), "We need to be on a boundary face, duh");
+  mooseAssert(!isInternalFace(fi), "A boundary face value has been requested on an internal face.");
```
Haha I do appreciate the added friendliness :smile:

lindsayad commented on modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:
```diff
@@ -84,7 +85,7 @@ InternalVolumetricFlowRate::computeQpIntegral()
+    // Get face value for velocity
     // FIXME Make sure getInternalFaceValue uses the right interpolation method, see #16585
     const auto & vx_face =
         !getFieldVar("vel_x", 0)
```
I want to make sure that we end up in a place where we don't have any `getFieldVar` calls in `*::computeQpIntegral`. I think here (and in a lot of other places), we can actually just do `_fv_vel_x ? do_if_non_null : do_if_null`

lindsayad replied:
(obviously replacing `_fv_vel_x` with the appropriate variable in other contexts)

lindsayad reviewed (changes requested):
Just the one comment!

lindsayad commented:
> As you can see in the csv, the inlet energy flow is still not '300' as it should be, but depends on the numerical scheme used. Further in the channel we get the right value, especially when the flow is close to being fully developed again. I expect this is due to not using the face value for the material property, the heat production is happening in the first cell too so the temperature has already increased.

Can you paste in a line plot of the temperature down the center of the channel?

GiudGiud committed:
Remove calls to getFieldVar by testing pointer to fv variables instead

GiudGiud commented:
Ok moved it away. Good call on using the pointer.
Profiles attached with upwind interpolation and with average interpolation
![fine_mesh_T](https://user-images.githubusercontent.com/25037151/106949085-d41fae00-672c-11eb-8fc4-7c21c0544a22.png)
![coarse_mesh_T](https://user-images.githubusercontent.com/25037151/106949086-d4b84480-672c-11eb-98b5-d81fae69226b.png)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/673621/) on 9643678 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16767/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16767/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b28b88dfcc3512ec6b067f428c14e805bcf28a1c`

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/gold/diverging_INSFV_quad_xy_upwind.csv:
```diff
@@ -0,0 +1,3 @@
+time,inlet_advected_energy,inlet_mass_constant,inlet_mass_matprop,inlet_mass_variable,inlet_momentum_x,inlet_momentum_y,mid1_advected_energy,mid1_mass,mid1_momentum_x,mid1_momentum_y,mid2_advected_energy,mid2_mass,mid2_momentum_x,mid2_momentum_y,outlet_advected_energy,outlet_mass,outlet_momentum_x,outlet_momentum_y
+0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
+1,-302,-1,-1,-1,-0.024476040741523,-0.94215502182725,310,1,-0.059947284338798,1.0723266687159,310,1,0.029410484806732,0.78663499152553,310,1,0.0034392597619175,0.76467916759833
```
I don't understand why `inlet_advected_energy` matches the cell centroid value adjacent to the boundary. I would have thought that the number would be lower because I would think that we are using the two term boundary expansion to extrapolate to the boundary face...and then the upwind ghost value would consequently be lower than the boundary face value because of the gradient.

GiudGiud replied:
it's a material property here.
With a variable you would be right

lindsayad replied:
Oh em gee, I hate the material system more every day. Yea this inherits from `SideFluxIntegral` which is not a `TwoMaterialPropertyInterface` so we only have access to the material property on one side of the face...which makes perfect sense for FE calculations, but doesn't give us the flexibility I would like for finite volume calculations. Yea you have `getADMaterialProperty` which will get the material property at the element centroid, when what we really want for an upwind scheme is `getNeighborADMaterialProperty` which would give us the value on the ghost cell. Could you add this to the issue where we document the need for a more flexible material system? The ticket number is #16809

lindsayad commented on modules/navier_stokes/src/postprocessors/VolumetricFlowRate.C:
```diff
@@ -16,23 +17,122 @@ InputParameters
+    _fv_vel_x(dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar("vel_x", 0))),
+    _fv_vel_y(dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar("vel_y", 0))),
+    _fv_vel_z(dynamic_cast<const MooseVariableFV<Real> *>(getFieldVar("vel_z", 0))),
+    _advected_variable_supplied(parameters.isParamSetByUser("advected_variable")),
```
I'm not going to make you change it, but in the future I believe we have a `MooseObject::isParamSetByUser` so you don't have to call through the `parameters` object

GiudGiud replied:
ok I ll catch it in PINSFV then

lindsayad reviewed (approved):
I'm really happy to have more framework and `navier_stokes` developers!
